File: BinaryTrees/19/validateBst.js
/**
 * Definition for a binary tree node.
 * function TreeNode(val, left, right) {
 *     this.val = (val===undefined ? 0 : val)
 *     this.left = (left===undefined ? null : left)
 *     this.right = (right===undefined ? null : right)
 * }
 */
/**
 * @param {TreeNode} root
 * @return {boolean}
 */
let lastSeenValue = null;
let valid = true;

var isValidBST = function (root) {
  inOrder(root);
  return valid;
};

var inOrder = function (root, prev) {
  if (root && valid) {
    inOrder(root.left);
    if (lastSeenValue !== null && root.val <= lastSeenValue) {
      valid = false;
      return;
    }
    lastSeenValue = root.val;
    if (valid) inOrder(root.right);
  }
};

File: BinaryTrees/19/validate_bst.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def __init__(self):
        self.lastSeenValue = None
        self.valid = True

    def isValidBST(self, root: TreeNode) -> bool:
        self.inOrder(root)
        return self.valid


    def inOrder(self, root):
        if root and self.valid:
            self.inOrder(root.left)
            if self.lastSeenValue is not None and root.val <= self.lastSeenValue:
                self.valid = False
                return
            self.lastSeenValue = root.val
            if self.valid:
                self.inOrder(root.right)